README.md
# a4adiags: Diagnostics for FLa4a stock Assessment Models

[![License](https://flr-project.org/img/eupl12.svg)](https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12)
[![R-CMD-check](https://github.com/flr/a4adiags/workflows/R-CMD-check/badge.svg)](https://github.com/flr/a4adiags/actions)

Additional Diagnostics for FLa4a stock Assessment Models

## Installation

Install the development version from GitHub with:

    devtools::install_github("flr/a4adiags")

This package depends on [FLa4a](https://flr-project.org/FLa4a), [FLasher](https://flr-project.org/FLasher) and [ggplotFL](https://flr-project.org/ggplotFL), available at the [FLR Project repository](https://flr-project.org/R) using

    install.packages(c("FLa4a", "FLasher", "ggplotFL"), repos="https://flr-project.org/R")

and *ggplot2*, *snpar*, *foreach* and *data.table*, available from CRAN

    install.packages(c("ggplot2", "snpar", "foreach", "data.table"))

To make use of the parallelization capabilities in the package, install *doParallel* or any other *foreach* backend.

## Usage

### Cross-validation with restrospective hindcasting

![](man/figures/README-xval.png)<!-- -->

### Runs tests

![](man/figures/README-runstest.png)<!-- -->

## Documentation

- [Help pages](http://flr-project.org/a4adiags)

## License

Copyright (c) 2020 European Union. Released under the [EUPL 1.2](https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12).

## Authors

Iago MOSQUEIRA (WMR), Henning WINKER (EC JRC).

## Contact

You are welcome to:

- Submit suggestions and bug-reports at: <https://github.com/flr/a4adiags/issues>
- Send a pull request on: <https://github.com/flr/a4adiags/>


NEWS
 a4adiags 0.2.0

 BUG FIXES

- plotRunsTest now does the right calculations for age-disaggregated indices.

 NEW FEATURES

- plotRunsTest(FLQuant, FLQuant)

 a4adiags 0.1.0

- First beta version, contains methods for runs test and cross-validation by
  retrospective hindcasting


tests/testthat/test-xval.R
# test-xval.R - DESC
# /test-xval.R

# Copyright Iago MOSQUEIRA (WMR), 2020
# Author: Iago MOSQUEIRA (WMR) <iago.mosqueira@wur.nl>
#
# Distributed under the terms of the EUPL-1.2


# LOAD data: FLStock + FLIndices

data("sol274")


# --- SET UP model

# fmodel, mimics AAP

fmod <- ~te(replace(age, age > 8, 8), year, k = c(4, 22)) +
  s(replace(age, age > 8, 8), k=4) +
  s(year, k=22, by=as.numeric(age==1))

# qmodel (GAM, SNS)

qmod <- list(~s(age, k=3), ~s(age, k=3))

# vmodel (catch, GAM, SNS)

vmod <- list(
  ~s(age, k=3),
  ~s(age, k=3),
  ~s(age, k=3))

# srmodel

srmod <- ~factor(year)

# fit

fit <- sca(stock, indices[c("BTS", "SNS")],
  srmodel=srmod, fmodel=fmod, qmodel=qmod, vmodel=vmod)

run <- stock + fit


# --- XVAL

library(doParallel)

if(Sys.info()["sysname"] %in% c("Darwin", "Linux")) {
  registerDoParallel(3)
} else {
  cl <- makeCluster(3)
  clusterEvalQ(cl = cl, expr = { library(FLa4a); library(FLasher) })
  registerDoParallel(cl)
}

# CALL xval

system.time(
res <- a4ahcxval(stock, indices, nyears=5, nsq=3, srmodel=srmod, fmodel=fmod,
  qmodel=qmod, vmodel=vmod)
)

# COMPUTE MASE, BUT dropping 2019 prediction from res$indices.

mase(indices, res$indices[-1])

# PLOT, drop observations index 1.

plotXval(indices, res$indices)


tests/testthat/test-runstest.R
# test-runstest.R - DESC
# a4adiags/test/testthat/test-runstest.R

# Copyright Iago MOSQUEIRA (WMR), 2020
# Author: Iago MOSQUEIRA (WMR) <iago.mosqueira@wur.nl>
#
# Distributed under the terms of the EUPL-1.2


# LOAD data: FLStock + FLIndices

data("sol274")


# --- SET UP model

# fmodel, mimics AAP

fmod <- ~te(replace(age, age > 8, 8), year, k = c(4, 22)) +
  s(replace(age, age > 8, 8), k=4) +
  s(year, k=22, by=as.numeric(age==1))

# qmodel (GAM, SNS)

qmod <- list(~s(age, k=3), ~s(age, k=3))

# vmodel (catch, GAM, SNS)

vmod <- list(
  ~s(age, k=3),
  ~s(age, k=3),
  ~s(age, k=3))

# srmodel

srmod <- ~factor(year)

# fit

fit <- sca(stock, indices[c("BTS", "SNS")],
  srmodel=srmod, fmodel=fmod, qmodel=qmod, vmodel=vmod)

# PLOT for a4aFitSA and FLIndices, by total numbers

plotRunstest(fit , indices, combine=TRUE) +
  ggtitle("SOL.27.4 - Total numbers")

# PLOT for a4aFitSA and FLIndices, by numbers at age

plotRunstest(fit , indices, combine=FALSE) +
  ggtitle("SOL.27.4 - Numbers by age")

# PLOT by total biomass

wts <- lapply(indices, catch.wt)

plotRunstest(index(fit) * wts, wts * lapply(indices, index)) +
  ggtitle("SOL.27.4 - Total biomass")

# CHANGE fill colors

plotRunstest(fit, indices) +
  scale_fill_manual(values=c("TRUE"="cyan", "FALSE"="pink"))

# CHANGE point colors

plotRunstest(fit, indices) +
    scale_colour_manual(values=c("FALSE"="green", "TRUE"="red"))

# CALL on single index

plotRunstest(index(fit)$SNS, index(indices$SNS), combine=FALSE)
plotRunstest(index(fit)$SNS, index(indices$SNS), combine=TRUE)


R/data.R
# data.R - Example datasets for the a4adiags package
# a4adiags/R/data.R

# Copyright Iago MOSQUEIRA (WMR), 2020
# Author: Iago MOSQUEIRA (WMR) <iago.mosqueira@wur.nl>

#' Data from the 2020 ICES stock assessment of North Sea sole (sol.27.4)
#' 
#' Aliquam sagittis feugiat felis eget consequat. Praesent eleifend dolor massa, 
#' vitae faucibus justo lacinia a. Cras sed erat et magna pharetra bibendum quis in 
#' mi. Sed sodales mollis arcu, sit amet venenatis lorem fringilla vel. Vivamus vitae 
#' ipsum sem. Donec malesuada purus at libero bibendum accumsan. Donec ipsum sapien, 
#' feugiat blandit arcu in, dapibus dictum felis. 
#' 
#' fmod <- ~te(replace(age, age > 8, 8), year, k = c(4, 22)) +
#'   s(replace(age, age > 8, 8), k=4) +
#'   s(year, k=22, by=as.numeric(age==1))
#' 
#' qmod <- list(~s(age, k=3), ~s(age, k=3))
#' 
#' vmod <- list(
#'   ~s(age, k=3),
#'   ~s(age, k=3),
#'   ~s(age, k=3))
#' 
#' srmod <- ~factor(year)
#' 
#' @docType data
#' @keywords datasets
#' @aliases stock indices fit
#' @format Objects of class FLStock, FLIndices and a4aFitSA
#' @name sol274
#' @rdname data
NULL


R/runstest.R
# runstest.R - DESC
# a4adiags/R/runstest.R

# Copyright Iago MOSQUEIRA (WMR), 2020
# Author: Iago MOSQUEIRA (WMR) <iago.mosqueira@wur.nl>
#
# Distributed under the terms of the EUPL-1.2

globalVariables(c("pass", "p.value", "qname", "age", "lcl", "ucl", "outlier"))

# sigma3 (FLQuant) {{{

#' Compute the 3-sigma limits and the corresponding p-value
#'
#' @param x An object of class FLQuant.
#' @param mixing Alternative hypothesis to be tested by snpar::runs.test. One of "two.sided", "less" (default) or "greater".
#'
#' @return A list with elements 'lcl', 'ucl' and 'p.value'.
#'
#' @examples
#' data(sol274)
#' sigma3(catch(stock))

sigma3 <- function(x, mixing="less") {

  # COMPUTE average moving rate

  mr <- abs(diff(x - 0))
  amr <- mean(mr, na.rm=TRUE)
  
  # COMPUTE upper limit for moving ranges

  ulmr <- 3.267 * amr

  # REMOVE moving ranges greater than ulmr and recalculate amr, Nelson 1982

  mr  <- c(mr)[c(mr) < ulmr]
  amr <- mean(mr, na.rm = TRUE)

  # Calculate standard deviation, Montgomery, 6.33

  stdev <- amr / 1.128

  # Calculate control limits
  lcl <- -3 * stdev
  ucl <- 3 * stdev

  if(nlevels(factor(sign(x))) > 1) {

    # Make the runs test non-parametric
    y = ifelse(x < 0, -1, 1)
    runstest = runs.test(c(y), alternative = mixing)
    pvalue = round(runstest$p.value, 3)

  } else {
    pvalue = 0.001
  }
 
return(list(lcl = lcl, ucl = ucl, p.value = pvalue))

} # }}}

# runstest {{{

#' Computes Runs Test p-values
#'
#' @param fit The result of a model fit.
#' @param obs The observations used in the fit.
#' @param combine Should ages be combined by addition, defaults to TRUE.
#' @param ... Extra arguments.
#'
#' @return A list with runs 'p.values' and 'pass' 
#'
#' @examples
#' data(sol274)
#' # Call on a a4aFitSA object
#' runstest(fit, indices)
#' #
#' runstest(index(fit), lapply(indices, index))
    
setGeneric("runstest", function(fit, obs, ...)
  standardGeneric("runstest"))

setMethod("runstest", signature(fit="FLQuants", obs="missing"),
  function(fit, combine=TRUE) {

    # COMBINE
    if(combine) {
      fit <- lapply(fit, quantSums)
    }
    
    # RESIDUALS
    res <- fit
    
    # sigma3, by index
    if(combine) {
      s3s <- lapply(res, sigma3)
      # or index and age
    } else {
      s3s <- lapply(res, function(x) {
        rbindlist(lapply(divide(x, 1), sigma3), idcol="age")
      })
    }

    s3dat <- rbindlist(lapply(s3s, as.data.frame), idcol="qname")
    
    # p.value >= 0.05 -> TRUE, green
    s3dat[, pass:=p.value >= 0.05]
    return(s3dat)
  }
)

setMethod("runstest", signature(fit="FLQuants", obs="FLQuants"),
  function(fit, obs, combine=TRUE) {
    
    # COMBINE
    if(combine) {
      fit <- lapply(fit, quantSums)
      obs <- lapply(obs, quantSums)
    }
    
    # RESIDUALS
    res <- FLQuants(mapply(residuals, obs, fit, SIMPLIFY=FALSE))
    
    return(runstest(res, combine=combine))
  }
)

setMethod("runstest", signature(fit="a4aFitSA", obs="FLIndices"),
  function(fit, obs, combine=TRUE) {
    
    # EXTRACT index fit
    fit <- index(fit)
    
    # EXTRACT index observations
    obs <- lapply(obs[names(fit)], index)
    
    return(runstest(fit, obs, combine=combine))
  }
)

#' @rdname runstest

setMethod("runstest", signature(fit="FLQuant", obs="FLQuant"),
  function(fit, obs, combine=TRUE) {
    
    fit <- FLQuants(A=fit)
    obs <- FLQuants(A=obs)
    
    return(runstest(fit, obs, combine=combine))
  }
)
# }}}

# plotRunstest {{{

#' Plot the runs test result for one or more time series
#'
#' @param fit The result of a model fit.
#' @param obs The observations used in the fit.
#' @param combine Should ages be combined by addition, defaults to TRUE.
#' @param ... Extra arguments.
#'
#' @return An object of class ggplot2::gg
#'
#' @examples
#' data(sol274)
#' plotRunstest(index(fit), lapply(indices, index))

setGeneric("plotRunstest", function(fit, obs, ...)
		standardGeneric("plotRunstest"))

#' @rdname plotRunstest

setMethod("plotRunstest", signature(fit="FLQuants", obs="missing"),
  function(fit, combine=TRUE) {

  # COMBINE
  if(combine) {
    fit <- lapply(fit, quantSums)
  }

  # RESIDUALS
  res <- fit
  
  # CREATE data.frame
  dat <- data.table(as.data.frame(res))

  # sigma3, by index
  s3dat <- runstest(fit)

  # FIND single limits for all indices
  lims <- c(min=min(unlist(lapply(res, dims, c("minyear")))),
    max=max(unlist(lapply(res, dims, c("maxyear")))))
  
  # MERGE s3dat into dat
  if(combine)
    dat <- merge(dat, s3dat[, list(qname, lcl, ucl, pass)], by=c('qname'))
  else {
    dat <- merge(dat, s3dat[, list(age, qname, lcl, ucl, pass)], by=c('qname', 'age'))
  }

  # ADD limits to colour outliers
  dat[, outlier:=data < lcl | data > ucl]
  
  # PLOT
  p <- ggplot(dat) +
    geom_rect(data=s3dat, aes(xmin=lims[1] - 1, xmax=lims[2] + 1,
      ymin=lcl, ymax=ucl, fill=pass), alpha=0.8) +
    scale_fill_manual(values=c("TRUE"="#cbe368", "FALSE"="#ef8575")) +
    geom_hline(yintercept=0, linetype=2) +
    geom_segment(aes(x=year, y=0, xend=year, yend=data), na.rm=TRUE) +
    geom_point(aes(x=year, y=data), size=1.5, na.rm=TRUE) +
    geom_point(aes(x=year, y=data, colour=outlier), size=1, na.rm=TRUE) +
    scale_colour_manual(values=c("FALSE"="#ffffff", "TRUE"="#d50102")) +
    xlab("") + ylab("Residuals") +
    theme(legend.position="none")

  if(combine)
    p <- p + facet_grid(qname ~ .)
  else
    p <- p + facet_grid(factor(age, levels=order(unique(age))) ~ qname,
      scales="free_y")

  return(p)
  }
)

setMethod("plotRunstest", signature(fit="FLQuants", obs="FLQuants"),
  function(fit, obs, combine=TRUE) {
  
  # COMBINE
  if(combine) {
    fit <- lapply(fit, quantSums)
    obs <- lapply(obs, quantSums)
  }

  # RESIDUALS
  res <- FLQuants(mapply(residuals, obs, fit, SIMPLIFY=FALSE))

  return(plotRunstest(res, combine=combine))
  }
)

#' @rdname plotRunstest

setMethod("plotRunstest", signature(fit="a4aFitSA", obs="FLIndices"),
  function(fit, obs, combine=TRUE) {

    # EXTRACT index fit
    fit <- index(fit)

    # EXTRACT index observations
    obs <- lapply(obs[names(fit)], index)

    plotRunstest(fit, obs, combine=combine)
  }
)

#' @rdname plotRunstest

setMethod("plotRunstest", signature(fit="FLQuant", obs="FLQuant"),
  function(fit, obs, combine=TRUE) {

    fit <- FLQuants(A=fit)
    obs <- FLQuants(A=obs)

    plotRunstest(fit, obs, combine=combine) +
      theme(strip.text = element_blank(), strip.background = element_blank())
  }
)
# }}}


demo/retro.R
# retro.R - DESC
# /retro.R

# Copyright Iago MOSQUEIRA (WMR), 2020
# Author: Iago MOSQUEIRA (WMR) <iago.mosqueira@wur.nl>
#
# Distributed under the terms of the EUPL-1.2


data(sol274)

# WINDOW stocks and indices

stks <- lapply(setNames(nm=seq(2019, length=6, by=-1)), function(x)
  window(stock, end=x))

idxs <- lapply(seq(2019, length=6, by=-1), function(x)
  window(indices[c("BTS", "SNS")], end=x))

# SUBMODELS

# fmodel
fmod <- ~te(replace(age, age > 8, 8), year, k = c(4, 22)) +
  s(replace(age, age > 8, 8), k=4) +
  s(year, k=22, by=as.numeric(age==1))

# qmodel (GAM, SNS)

qmod <- list(~s(age, k=3), ~s(age, k=3))

# vmodel (catch, GAM, SNS)

vmod <- list(
  ~s(age, k=3),
  ~s(age, k=3),
  ~s(age, k=3))

# srmodel

srmod <- ~factor(year)

# RUN sca
retrf <- mapply(function(...) do.call(sca, list(...)),
  stock=stks, indices=idxs, srmodel=list(srmod), fmodel=list(fmod),
  qmodel=list(qmod), vmodel=list(vmod), SIMPLIFY=FALSE)

# UPDATE stocks
retrs <- FLStocks(mapply("+", stks, retrf))

# COMPUTE mohnMatrix

mmats <- lapply(setNames(nm=c("fbar", "ssb", "rec")), mohnMatrix, stocks=retrs)

# CALCULATE mohn's rho

mrhos <- lapply(mmats, icesAdvice::mohn)


data-raw/sol274.R
# sol274.R - DESC
# /sol274.R

# Copyright Iago MOSQUEIRA (WMR), 2020
# Author: Iago MOSQUEIRA (WMR) <iago.mosqueira@wur.nl>
#
# Distributed under the terms of the EUPL-1.2


library(FLa4a)

# ORIGINAl objexcts from https://github.com/ices-taf/2020_sol.27.4_assessment

load("sol274.RData")

# fmodel, mimics AAP

fmod <- ~te(replace(age, age > 8, 8), year, k = c(4, 22)) +
  s(replace(age, age > 8, 8), k=4) +
  s(year, k=22, by=as.numeric(age==1))

# qmodel (GAM, SNS)

qmod <- list(~s(age, k=3), ~s(age, k=3))

# vmodel (catch, GAM, SNS)

vmod <- list(
  ~s(age, k=3),
  ~s(age, k=3),
  ~s(age, k=3))

# srmodel

srmod <- ~factor(year)

# fit

fit <- sca(stock, indices[c("BTS", "SNS")],
  srmodel=srmod, fmodel=fmod, qmodel=qmod, vmodel=vmod)

save(fit, stock, indices, file="../data/sol274.RData", compress="xz")


R/xval.R
# xval.R - DESC
# a4adiags/R/xval.R

# Copyright Iago MOSQUEIRA (WMR), 2020
# Author: Iago MOSQUEIRA (WMR) <iago.mosqueira@wur.nl>
#
# Distributed under the terms of the EUPL-1.2

globalVariables(c("final", "y", "pred"))

# a4ahcxval {{{

#' Compute a retrospective hindcast cross-validation of a4a stock and indices
#'
#' @param stock Input FLStock object.
#' @param indices Input FLIndices object.
#' @param nyears Number if years for retrospective, defaults to 5.
#' @param nsq Number of years for average biology and selectivity, defaults to 3.
#' @param fixed.ks Is the number of knots is splines with 'year' constant?
#' @param ... Any submodels and other arguments for the call to *sca*.
#'
#' @return A list containing elements 'stock', of class *FLStocks*, and
#' 'indices', a list of *FLIndices* objects.

a4ahcxval <- function(stock, indices, nyears=5, nsq=3, fixed.ks=FALSE, ...) {

  fy <- dims(stock)$maxyear
  y0 <- dims(stock)$minyear

  # CHECK submodels
  mods <- list(...)

  foo <- function(x)
    sum(grepl("te\\(*", labels(terms(x))) & grepl("year", labels(terms(x))) |
    grepl("s\\(*", labels(terms(x))) & grepl("year", labels(terms(x))))

  smyr <- lapply(mods, function(x){
    if(is(x, "list"))
      unlist(lapply(x, foo))
    else
      foo(x)
  })

  if(any(unlist(smyr) > 0))
    warning("Submodels using s(year) or te(year) should have 'k' changed in retro")

  # SELECT indices that fall within retro year range
  iyrs <- unlist(lapply(indices, function(x) dims(x)$maxyear)) >= (fy - nyears)

  if(!all(iyrs))
    warning(paste("FLIndices outside of year range are excluded from xval:",
      paste(names(indices)[!iyrs], collapse=", ")))

  # FILL index wt using stock.wt
  indices <- lapply(indices, function(x) {
    dmns <- dimnames(index(x))
    if(all(is.na(catch.wt(x))))
      catch.wt(x) <- stock.wt(stock)[dmns$age, dmns$year]
    return(x)
    }
  )

  # LOOP

  retro <- foreach(y=seq(fy, fy - nyears)) %dopar% {

    # RUN
    fit <- sca(window(stock, end=y), lapply(indices, window, end=y), ...)

    # UPDATE
    stock.n(stock)[, ac(y0:y)] <- stock.n(fit) 
    harvest(stock)[, ac(y0:y)] <- harvest(fit)

    # PREDICT stock, unless y == fy
    if(y < fy) {

     # SET future SELEX, WTs
      fyrs <- ac(seq(y-nsq-1, y))

      harvest(stock)[, ac(y+1)] <- yearMeans(harvest(stock)[, fyrs])
      stock.wt(stock)[, ac(y+1)] <- yearMeans(stock.wt(stock)[, fyrs])
      catch.wt(stock)[, ac(y+1)] <- yearMeans(catch.wt(stock)[, fyrs])
      landings.wt(stock)[, ac(y+1)] <- yearMeans(landings.wt(stock)[, fyrs])
      discards.wt(stock)[, ac(y+1)] <- yearMeans(discards.wt(stock)[, fyrs])

      # REC
      srr <- predictModel(model=rec~a,
        params=FLPar(a=exp(mean(log(rec(stock))))))

      # FWD
      pred <- fwd(stock, control=fwdControl(year=seq(y+1, fy),
        quant="catch", value=catch(stock)[, ac(seq(y+1, fy))]), sr=srr)
    } else {
      pred <- stock
    }

    # PREDICT FLIndices

    qs <- predict(fit)$qmodel

    ihat <- mapply(function(a, b) {
     
      # GET dims
      dmns <- dimnames(a)
      dis <- dims(a)
      timf <- mean(range(b)[c("startf", "endf")])

      # UPDATE index
      a <- window(a, end=fy)

      # EXTEND cachabilities if empty
      if(y < fy) {
        a[, ac(seq(dis$maxyear + 1, fy))] <-
          yearMeans(a[, ac(seq(dis$maxyear - nsq + 1, dis$maxyear))])
      }
      
      # COMPUTE predicted index
      index(b) <- a * stock.n(pred)[dmns$age, ac(seq(dis$minyear, fy))] *
        exp(-z(pred)[dmns$age, ac(seq(dis$minyear, fy))] * timf)

      # STORE catchabilities
      index.q(b) <- a

      return(b)

      }, a=qs[iyrs], b=indices[iyrs], SIMPLIFY=FALSE)

    name(pred) <- paste(name(pred), y, sep="_")
    desc(pred) <- paste(".", desc(pred), "xval -", y)

    list(stock=pred, indices=ihat)
  }

  # OUTPUT: stocks (FLStocks)
  stocks <- FLStocks(lapply(retro, "[[", "stock"))
  names(stocks) <- seq(fy, fy - nyears)

  # indices (list of FLIndices))
  indices <- lapply(retro, function(x) FLIndices(x$indices))
  names(indices) <- seq(fy, fy - nyears)

  list(stocks=stocks, indices=indices)
} # }}}

# mase {{{

#' Compute mean absolute scaled error (MASE)
#'
#' Franses, PH. "A note on the Mean Absolute Scaled Error". International Journal of Forecasting. 32 (1): 20–22. doi:10.1016/j.ijforecast.2015.03.008.
#'
#' @param ref Reference or naive prediction.
#' @param preds Predicitions to compare to reference.
#' @param order Are predictions in 'inverse' (default) or 'ahead' order.
#' @param ... Extra arguments.
#'
#' @return A numeric vector of the same length as 'preds'.

setGeneric("mase", function(ref, preds, ...) standardGeneric('mase'))

#' @rdname mase

setMethod("mase", signature(ref="FLQuant", preds="FLQuants"),
  function(ref, preds, order=c("inverse", "ahead")) {

    # SET dims
    fy <- dims(ref)$maxyear
    nyears <- length(preds)

    # REVERSE if ahead
    if(order[1] == "ahead")
      preds <- preds[rev(seq(length(preds)))]

    # ADD names if missing
    if(is.null(names(preds)))
      names(preds) <- seq(fy - 1, fy - nyears)

    # \sum_{t=T-h+1}^{T} |\hat{y}_t - y_t|
    num  <- abs(log(mapply(function(x, y) x[, y], preds,
      y=ac(seq(fy, fy - nyears + 1)), SIMPLIFY=TRUE)) -
      log(ref[, ac(seq(fy, fy - nyears + 1))]))

    # \sum_{t=T-h+1}^{T} |y_t - y_{t-1}|
    den <- abs(log(ref[, ac(seq(fy, fy - nyears + 1))]) -
      log(ref[, ac(seq(fy - 1, fy - nyears))]))

    mase <- (1 / nyears * sum(num)) / (1 / nyears * sum(den))

    return(mase)
  }
)

#' @rdname mase
#' @param wt Mean weights-at-age to use with indices.

setMethod("mase", signature(ref="FLIndices", preds="list"),
  function(ref, preds, order="inverse", wt="missing") {

    # CHECK classes in list
    if(!all(unlist(lapply(preds, is, "FLIndices"))))
      stop("All elements in 'preds' list must be of class 'FLIndices'.")

    if(!all(unlist(lapply(preds, length)) == length(ref)))
      stop("'FLIndices' in 'preds' must have the same length as 'ref'.")

    # TODO CHECK names and warn if first matches last year of ref

    indices <- c(list(ref), preds)

    # TODO PARSE wt and add to indices

    res <- unlist(lapply(setNames(nm=names(indices[[1]])), function(i) {
      # COMPUTE index in biomass
      flqs <- lapply(indices, function(x) {
        if(is(x, "FLIndexBiomass"))
          index(x[[i]])
        else
          quantSums(index(x[[i]]) * catch.wt(x[[i]]))
      })
      mase(flqs[[1]], FLQuants(flqs[-1]), order=order)
    }))

    return(res)
  }
)
# }}}

# data.tables {{{

dtp <- function(flis, y0) {

  rbindlist(lapply(names(flis), function(i) {
    pred <- pmin(an(i) + 1, an(names(flis)[1]))
    data.table(cbind(as.data.frame(lapply(flis[[ac(i)]], function(x) {
      # COMPUTE Total abundance in biomass
      window(quantSums(index(x) * catch.wt(x)), start=y0, end=pred)
    }), drop=TRUE, qname="index"), final=i, pred=pred))}))
}

dto <- function(flis, y0) {

  data.table(as.data.frame(lapply(flis, function(x) {
    dmns <- dimnames(x)
    window(quantSums(index (x) * catch.wt(x)), start=y0)
  }), drop=TRUE, qname="index"))

}
# }}}

# plotXval (FLIndices, list) {{{

#' Plot of FLIndices cross-validation by retrospective hindcast
#'
#' @param x An *FLIndices* object of the original observations.
#' @param y A list contaning *FLIndices* objects returned by *a4ahcxval*.
#' @param order Order in which retrospective runs are stored, defaults to"inverse".
#'
#' @return A ggplot object
#'
#' @examples
#' # SEE vignette

plotXval <- function(x, y, order="inverse") {

  # SET first year of plot as 1st of retro - 3
  y0 <- dims(x[[1]])$maxyear - length(y) - 2
  py <- do.call(seq, as.list(rev(dims(x[[1]])$maxyear - c(0, length(y)  - 2))))

  # CONVERT inputs to data.tables
  dato <- dto(x, y0)
  datp <- dtp(y, y0)

  # CALCULATE mase
  if(order == "inverse")
    idr <- 1
  else if (order == "ahead")
    idr <- length(y)
  else {
    idx <- an(names(y)) == dims(x[[1]])$maxyear
    if(sum(idx) == 1)
      idr <- which(idr)
    else
      stop("Could not identify reference run (to last year).")
  }

  # CALCULATE mase
  imase <- mase(x, y[-idr], order=order)

  # GENERATE facet labels
  lbs <- unlist(lapply(seq(length(imase)), function(x)
    paste(names(imase)[x], "\nMASE:", format(imase[x], digits=3))))
  names(lbs) <- names(imase)

  llb <- names(y)
  llb[idr] <- paste(llb[idr], "(ref)")
  # PLOT

  p <- ggplot(datp, aes(x=year, y=data, colour=final)) +

  # data lines and points
  geom_line(data=dato, linetype=2, colour="gray") +
  geom_point(data=dato, colour="black", size=3) +
  geom_point(data=dato, colour="white", size=2.6) +
  geom_point(data=dato[year %in% py,], aes(colour=ac(year-1)), size=2.6) +
  
  # retro lines and hindcast point
  geom_line() + geom_point(data=datp[year==pred, ]) +

  # format
  facet_wrap(~index, scales="free_y", ncol=2, labeller=as_labeller(lbs)) +
  xlab("") + ylab("") + scale_color_manual("", labels=rev(llb),
    values=rev(c("#000000", "#E69F00", "#56B4E9", "#009E73", "#D55E00",
    "#E69F00", "#56B4E9", "#009E73", "#D55E00", "#0072B2"))) +
  theme(legend.position="bottom")

  return(p)
}
# }}}
